homeassistant/components/nanoleaf/config_flow.py

```diff
@@ -0,0 +1,227 @@
+from homeassistant.const import CONF_HOST, CONF_TOKEN
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers.typing import DiscoveryInfoType
+from homeassistant.util.json import load_json, save_json
+
+from .const import DOMAIN
+from .util import pynanoleaf_get_info
+
+_LOGGER = logging.getLogger(__name__)
+
+# For discovery integration import
+CONFIG_FILE: Final = ".nanoleaf.conf"
+
+USER_SCHEMA: Final = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+    }
+)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Nanoleaf config flow."""
+
+    VERSION = 1
+
+    nanoleaf: Nanoleaf
+
+    # For discovery integration import
+    discovery_conf: dict
+    device_id: str
+
+    def __init__(self) -> None:
+        """Initialize a Nanoleaf flow."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle Nanoleaf flow initiated by the user."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=USER_SCHEMA, last_step=False
+            )
+        self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
+        self.nanoleaf = Nanoleaf(user_input[CONF_HOST])
+        try:
+            await self.hass.async_add_executor_job(self.nanoleaf.authorize)
+        except Unavailable:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=USER_SCHEMA,
+                errors={"base": "cannot_connect"},
+                last_step=False,
+            )
+        except NotAuthorizingNewTokens:
+            pass
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unknown error connecting to Nanoleaf")
+            return self.async_show_form(
+                step_id="user",
+                data_schema=USER_SCHEMA,
+                last_step=False,
+                errors={"base": "unknown"},
+            )
+        return self.async_show_form(step_id="link")
+
+    async def async_step_zeroconf(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf Zeroconf discovery."""
+        _LOGGER.debug("Zeroconf discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def async_step_homekit(self, discovery_info: DiscoveryInfoType) -> FlowResult:
+        """Handle Nanoleaf Homekit discovery."""
+        _LOGGER.debug("Homekit discovered: %s", discovery_info)
+        return await self._async_discovery_handler(discovery_info)
+
+    async def _async_discovery_handler(
+        self, discovery_info: DiscoveryInfoType
+    ) -> FlowResult:
+        """Handle Nanoleaf discovery."""
+        host = discovery_info["host"]
+        name = discovery_info["name"].replace(f".{discovery_info['type']}", "")
+        await self.async_set_unique_id(name)
```

Comment:
Can we get the bulb serial number here instead?

Reply:
> Can we get the bulb serial number here instead?

No, we need a token to get the serial number data.

The name is unique, printed on the device and cannot be changed.